blog/_src/posts/2017-11-XX-adding-refinement-types.md

```diff
@@ -0,0 +1,449 @@
+constraints.
+
+## Dependent function types
+
+By combining refinement types with dependent function types
+(i.e. functions whose domain and range may depend on argument and/or
+program values), we can now give a more precise function type for many
+basic vector operations, e.g.:
+
+```racket
+(: safe-vector-ref 
+   (All (A) (-> ([v : (Vectorof A)]
+                 [n : (v) (Refine [i : Integer]
+                                  (<= 0 i (- (vector-length v) 1)))])
+                A)))
+(define safe-vector-ref vector-ref)
+```
+
+Here we have stated that `safe-vector-ref`'s second argument (`n`)
+depends on its first argument (`v`) and that the type of `n` is an
+`Integer` which is a valid index for `v`.
+
+If we try and use `safe-vector-ref` with a vector and index which
+Typed Racket cannot prove satisfy the respective types:
+
+```racket
+(safe-vector-ref (vector 0 1) 2)
+```
+
+Typed Racket will report an error while type checking:
+
+```
+Type Checker: type mismatch
+  expected: (Refine
+             (i : Integer)
+             (and (<= i (- (vector-length v) 1)) (<= 0 v)))
+  given: (Refine (z : Positive-Byte) (= 2 z)) in: 2
+```
+
+### Preconditions
+
+Sometimes the refinements used in a dependent function type can also
+be written conveniently as a "precondition", e.g.:
+
+```racket
+(: safe-vector-ref
+   (All (A) (-> ([v : (Vectorof A)]
+                 [n : Integer])
+                #:pre (v n) (<= 0 n (- (vector-length v) 1))
+                A)))
+(define safe-vector-ref vector-ref)
+```
+
+This function type is equivalent to the previous version, but the
+refining proposition that requires that `n` be a valid index for `v`
+now directly references `n` and has been written directly after the
+list of arguments:
+
+```racket
+#:pre (v n) (<= 0 n (- (vector-length v) 1))
+```
+
+We can read the precondition specification above as follows: the
+precondition for the function depends on arguments `v` and `n` and it
+requires that Typed Racket can prove `(<= 0 n (- (vector-length v)
+1))` would produce `#true` at all call sites.
+
+Typed Racket may be able to provide more detailed feedback when the
+arguments are of the correct type but the precondition cannot be
+proved:
+
+```
+Type Checker: could not apply function;
+ unable to prove precondition
+  precondition: (<= 2 (- (vector-length v) 1)) 
+  in: (safe-vector-ref (vector 0 1) 2)
+```
+
+Here, Typed Racket first performed type inference and confirmed that
+the arguments `(vector 0 1)` and `2` had acceptable types before
+discovering the precondition was unprovable at the call site.
+
+We believe the error messages produced when using preconditions may be
+more informative in general. Your mileage may vary.
+
+## Programming with refinement types
+
+_Note: these examples should work on any Racket builds from 3 November
+2017 or later.  To use with Racket v6.11, an additional small snippet
+of code (included at the end of this post) is required._
```

Comment:
This text should link to that part of the post.

I think you'll want to do that with HTML:
https://stackoverflow.com/a/6494918/5237018